guides/effectsize_exemplar_within.Rmd

```diff
@@ -31,93 +31,266 @@ format_num <- function(nums, sigdigits = 3) gsub("\\.$", "", formatC(nums, sigdi
+    return()
+}
+```
+
+#### Iterate
+
+The bootstrap needs to run many times to determine how a subset of the data impacts
+
+```{r within-bootstrap-bootstrap iterator}
+
+# run many iterations of the bootstrap procedure
+analyzeManyIterations = function (bootstrapIterationCount) {
+  # each core needs to reload the libraries
+  library(tidyverse)
+  library(afex)
+  lapply(1:bootstrapIterationCount, function(x) analyzeIteration(x)) %>% 
+    bind_rows()
+}
+
+# data used for bootstrapping will collapse by each subject-condition combination
+dataAggregated <- data %>% 
+  group_by(layout, size, color, subject) %>% 
+  summarize(response_time = mean(response_time))
+
 ```
 
-*Note that the fractional degrees of freedom result from a Greenhousse-Geisser sphericity correction.*
 
+#### Parallelize
+
+Bootstrapping can be slow, especially with thousands of iterations. Splitting the iterations across processor cores cuts down on the wait time.
+
+```{r within-parallelize-boostrap, cache=TRUE}
+BOOTSTRAP_COUNT <- 100 # at least 5k recommended. Use lower values for quicker testing
+
+# Initiate cluster
+coreCount <- detectCores() - 1
+ifelse(coreCount < 1, 1, coreCount)
+cl <- makeCluster(coreCount)
+clusterExport(cl, "analyzeIteration")
+clusterExport(cl, "dataAggregated")
+# how many times should each core iterate 
+bootstrapIterationCount <- BOOTSTRAP_COUNT / coreCount
+# run the bootstrap and output the time
+system.time(
+  bootResults <- parLapply(cl, rep(bootstrapIterationCount, coreCount), analyzeManyIterations)
+)
+# the cluster is no longer needed
+stopCluster(cl)
+# cleanup
+rm(coreCount, cl, bootstrapIterationCount, dataAggregated)
 ```
-TODO: Boostrapped 95%CIs for effect sizes
-Pro: people should
-Con: would make the guide even longer
-Maybe push into another guideline?
+
+### Getting a confidence interval from a bootstrap
+
+Each bootstrap iterations ran one anaylsis, so wwe now have many results. So for each effect size, we sort the results and find the range of the inner 95%.
+
+```{r within-boostrap-3}
+# inner 95%
+PERCENTILE_LO <- 0.025
+PERCENTILE_HI <- 0.975
+
+# put all the boostraped results together
+bootResults <- bind_rows(bootResults)
+bootResults <- 
+  bootResults %>% 
+  group_by(effect) %>% 
+  summarize(
+    ESPercentileLo = unname(quantile(ges, probs = PERCENTILE_LO)),
+    ESPercentileHi = unname(quantile(ges, probs = PERCENTILE_HI)))
+
+anovaResults <- 
+  anovaResults %>% 
+  left_join(bootResults, by = 'effect')
+anovaResults %>% as.tibble()
+  
+ggplot(anovaResults) +
+  aes(y=forcats::fct_rev(forcats::fct_inorder(effect)), x=ges) +
+  geom_vline(xintercept = 0, linetype = 'dotted') +
+  geom_segment(aes(x=ESPercentileLo, xend=ESPercentileHi, yend=effect), size=1) +
+  geom_point(size=3) +
+  scale_x_continuous(expand = c(0,0)) + expand_limits(x = -0.05) +
+  expand_limits(x = max(anovaResults$ESPercentileHi) * 1.05) +
+  labs(x=expression(paste('Effect size ', eta[G]^2)), y='Effect') +
+  theme_bw()
```

Comment:
Here's a suggestion to simplify and separate concerns (manipulation from plotting):

```r
import::from(forcats, fct_rev, fct_inorder)

p_anova_results <- 
anovaResults %>% 
  
  # reverse order the factors so that they appear in proper order in the plot
  mutate(effect = fct_rev(fct_inorder(effect))) %>% 
  
  # plot and mapping
  ggplot(aes(x = effect, y = ges, ymin = ESPercentileLo, ymax = ESPercentileHi)) +
  
    # reference line of no-effect
    geom_hline(yintercept = 0, linetype = 'dotted') +
    
    # point- and interval-estimates
    geom_pointrange() +
  
    # ensures that we see the reference line
    expand_limits(x = 0) +
    
    # flip to horizontal plot and apply black-and-white theme
    coord_flip() +
    theme_bw()
```